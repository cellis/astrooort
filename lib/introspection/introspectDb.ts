import { Introspection } from '@wmfs';
import PgInfo from '@wmfs/pg-info';
import { Client } from 'pg';
import { loadIndexes, RawIndex } from './indexesIntrospection';

export interface IntrospectDB {
  indexes: RawIndex[];
  info: Introspection;
}

export default async (
  client: Client,
  schemas: string[]
): Promise<IntrospectDB> => {
  const info = await PgInfo({
    client,
    schemas,
  });

  const indexes = await loadIndexes(client, schemas.join(','));

  return {
    indexes,
    info,
  };
};
