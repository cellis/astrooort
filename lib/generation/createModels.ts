import { IntrospectDB } from '../introspection/introspectDb';
import createModel from './createModel';

export default function createModels(
  models: Record<string, Superluminal.Model>,
  introspection: IntrospectDB
) {
  const { info, indexes } = introspection;

  for (const [schemaName, schema] of Object.entries(info.schemas)) {
    for (const [tableName, table] of Object.entries(schema.tables)) {
      createModel(tableName, schemaName, table, models, indexes);
    }
  }

  return models;
}
