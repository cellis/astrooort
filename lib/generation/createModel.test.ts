// need a column serializer that takes column

import { Client } from 'pg';
import introspectDb from '../introspection/introspectDb';
import connectTestDb from '../test/connectTestDb';
import {
  userFullTextIdxName,
  userPrimaryKeyName,
} from '../test/fixtures/MockData';
import createModels from './createModels';

describe('createModel', () => {
  let client: Client;
  let indexes: Record<string, Superluminal.Index>;
  let User: Superluminal.Model;
  let Product: Superluminal.Model;

  beforeAll(async (done) => {
    client = await connectTestDb();

    const introspection = await introspectDb(client, ['superluminal']);
    const { models } = await createModels({}, introspection, {
      output: '.tmp',
    }, false);

    User = models.user;
    Product = models.product;
    indexes = User.indexes || {};

    done();
  });

  afterAll(() => {
    return client.end();
  });

  describe('table schema', () => {
    it('sets the schema on the model', () => {
      expect(User.schema).toBe('superluminal');
    });
  });

  describe('when a table has indexes', () => {
    it('generates a model with indexes', () => {
      expect(indexes).toMatchSnapshot();
      expect(indexes[userFullTextIdxName]).toEqual<Superluminal.Index>({
        columns: ['full_text'],
        unique: false,
      });
    });

    describe('a unique index', () => {
      it('generates a model with a unique index', () => {
        expect(indexes[userPrimaryKeyName]).toEqual<Superluminal.Index>({
          columns: ['slug'],
          unique: true,
        });
      });
    });
  });

  describe('when a table has columns', () => {
    it('handles default columns', () => {
      expect(User.columns['created_at'].default).toBe('now()');
    });

    it('generates a model with columns', () => {
      expect(User.columns['first_name']).toEqual<Superluminal.Column>({
        name: 'first_name',
        dataType: 'character varying',
        array: false,
        type: 'string',
        nullable: true,
        autoIncrement: false,
      });
    });

    it('handles autoincrementing column types, ok-ish', () => {
      expect(Product.columns['id']).toMatchObject<Partial<Superluminal.Column>>(
        {
          autoIncrement: true,
        }
      );
    });

    it('handles column primitive types properly', () => {
      expect(User.columns['credits'].type).toBe('number');
    });

    it('handles column array types properly', () => {
      expect(User.columns['hobbies'].type).toBe('Array<string>');
    });
  });
});
