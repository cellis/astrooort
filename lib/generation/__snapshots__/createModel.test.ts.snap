// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createModel when a table has indexes generates a model with indexes 1`] = `
Object {
  "user_full_text_idx": Object {
    "columns": Array [
      "full_text",
    ],
    "unique": false,
  },
  "user_pkey": Object {
    "columns": Array [
      "slug",
    ],
    "unique": true,
  },
}
`;
